reposerver/repository/repository.go

```diff
@@ -2659,3 +2661,103 @@ func (s *Service) GetGitDirectories(_ context.Context, request *apiclient.GitDir
 		Paths: paths,
 	}, nil
 }
+
+// UpdateRevisionForPaths compares two git revisions and checks if the files in the given paths have changed
+// If no files were changed, it will store the already cached manifest to the key corresponding to the old revision, avoiding an unnecessary generation.
+// Example: cache has key "a1a1a1" with manifest "x", and the files for that manifest have not changed,
+// "x" will be stored again with the new revision "b2b2b2".
+func (s *Service) UpdateRevisionForPaths(_ context.Context, request *apiclient.UpdateRevisionForPathsRequest) (*apiclient.UpdateRevisionForPathsResponse, error) {
+	repo := request.GetRepo()
+	revision := request.GetRevision()
+	syncedRevision := request.GetSyncedRevision()
+	refreshPaths := request.GetPaths()
+
+	if repo == nil {
+		return nil, status.Error(codes.InvalidArgument, "must pass a valid repo")
+	}
+
+	if len(refreshPaths) == 0 {
+		// Always refresh if path is not specified
+		return &apiclient.UpdateRevisionForPathsResponse{}, nil
+	}
+
+	gitClientOpts := git.WithCache(s.cache, true)
+	gitClient, revision, err := s.newClientResolveRevision(repo, revision, gitClientOpts)
```

Comment:
This will fail if the source is a non-git helm chart. 

IMO, if `source.IsHelm()`, then return early.  `syncedRevision` will be the chart revision and the "refs" revision that were used to generate it are not in the state. It makes it impossible to compare if a path of one of the refs has changed. Since Multi-source is in beta, I think it is worth opening an enhancement issue for this scenario.

manifest:
```yaml
  sources:
    - repoURL: https://helm.github.io/examples
      chart: hello-world
      targetRevision: 0.1.0
```

status:
```yaml
  sync:
    comparedTo:
      sources:
        - chart: hello-world
          helm:
            valueFiles:
              - $gitops-test/services/multi-source-ref/values.yaml
          path: whatever
          repoURL: https://helm.github.io/examples
          targetRevision: 0.1.0
        - path: services/multi-source-with-ref
          ref: gitops-test
          repoURL: https://github.com/agaudreault/gitops-tests
          targetRevision: test-sync
    revisions:
      - 0.1.0  # <---- syncedRevision, not a commit sha of https://helm.github.io/examples
      - aadddbe7f15bbcf800a91abd6ba5dc32dd8a63ad
```

Reply:
Good point, I added the check in the controller to avoid the call altogether when the source is Helm